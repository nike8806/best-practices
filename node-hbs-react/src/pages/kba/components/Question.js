import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Question extends Component {
  static propTypes = {
    type: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    answers: PropTypes.arrayOf(PropTypes.string).isRequired,
    onAnswered: PropTypes.func,
    disabled: PropTypes.bool
  };

  static defaultProps = {
    disabled: false,
    onAnswered: null
  };

  state = {
    // eslint-disable-next-line react/destructuring-assignment, react/no-unused-state
    type: this.props.type,
    answer: null
  };

  static getDerivedStateFromProps(props, state) {
    if (state.type !== props.type) {
      return {
        type: props.type,
        answer: null
      };
    }

    return null;
  }

  /**
   * Sync state with user's selection
   *
   * @param {SyntheticEvent} event The React `SyntheticEvent`
   */
  handleAnswer = (event) => {
    this.setState({ answer: event.target.value });
  };

  /**
   * Invoque prop callback and clear state when clicking Next
   */
  handleNextClick = () => {
    const { onAnswered, type } = this.props;

    if (onAnswered) {
      const { answer } = this.state;
      onAnswered({
        type,
        answer
      });
    }
  };

  /**
   * Render method
   *
   * @return {ReactElement}
   */
  render() {
    const {
      title, type, answers, disabled
    } = this.props;

    const { answer: currentAnswer } = this.state;
    return (
      <div className="question">
        <h5 className="question__title todo-heading">{title}</h5>

        <ul className="question__answers margin-bottom-32 list-unstyled">
          {answers.map(answer => (
            <li key={`${type}-${answer}`} className="question__answer">
              <label
                className="question__answer-text"
                htmlFor={`${type}-${answer}-id`}
              >
                <input
                  id={`${type}-${answer}-id`}
                  type="radio"
                  name="answer"
                  value={answer}
                  className="question__answer-selection todo-radio"
                  checked={currentAnswer === answer}
                  onChange={this.handleAnswer}
                  disabled={disabled}
                />
                <span className="question__answer-radio todo-radio-selector" />
                {answer}
              </label>
            </li>
          ))}
        </ul>

        <button
          type="button"
          className="btn btn-todo center-block question__next"
          disabled={disabled || !currentAnswer}
          onClick={this.handleNextClick}
        >
          Next
        </button>
      </div>
    );
  }
}

export default Question;
