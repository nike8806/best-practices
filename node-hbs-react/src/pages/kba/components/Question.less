.question__answer {
  list-style: none;
  border: 1px solid #eee;
  margin-bottom: 5px;
  background-color: #f5f5f5;
  padding: 0;

  input[type='radio']&-selection {
    margin-right: 8px;
    margin-top: 0;
    vertical-align: middle;

    width: 2em;
    margin: 0;
    padding: 0;
    font-size: 1em;
    opacity: 0;

    display: inline-block;
    margin-left: -2em;
    line-height: 1.5em;
  }
}

.question__answer-text {
  color: #6f7883;
  display: block;
  padding: 15px 15px 15px 10px;
  cursor: pointer;
}
