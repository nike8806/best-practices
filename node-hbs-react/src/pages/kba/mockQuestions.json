{
  "loanId": 127235022,
  "status": true,
  "idNumber": 1944906967,
  "questions": {
    "question": [
      {
        "type": "current.county.b",
        "prompt": "In which county have you lived?",
        "answer": [
          "FULTON",
          "APPANOOSE",
          "LAFAYETTE",
          "None of the above"
        ]
      },
      {
        "type": "previous.address",
        "prompt": "At which of the following addresses have you lived?",
        "answer": [
          "3 CRESSING CT",
          "16015 RACETRACK RD",
          "510 ADAMS RD",
          "None of the above"
        ]
      },
      {
        "type": "purchased.property.from",
        "prompt": "From whom did you purchase the property at 222333 PEACHTREE PLACE?",
        "answer": [
          "A VIRAY",
          "ERIC WALTORS",
          "JOE ANDERSON",
          "None of the above"
        ]
      }
    ]
  }
}
