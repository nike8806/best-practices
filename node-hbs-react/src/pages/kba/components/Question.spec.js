import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Question from './Question';
import mockQuestions from '../mockQuestions.json';

// TODO: Put this configuration in just one place
import Adapter from 'enzyme-adapter-react-16';
Enzyme.configure({ adapter: new Adapter() });

const question = mockQuestions.questions.question[0];

let component;
let handleAnswer;

describe('KBA Question component', () => {
  beforeEach(() => {
    handleAnswer = jest.fn();

    component = shallow(
      <Question type={question.type} title={question.prompt} answers={question.answer} onAnswered={handleAnswer} />
    );
  });

  it('should render correctly', () => {
    expect(component).toMatchSnapshot();
  });

  it('should disable Next button if no answer is selected', () => {
    expect(component.find('button.question__next').props().disabled).toBe(true);
  });

  it('should set state with the selected answer', () => {
    const randomOption = Math.floor(Math.random() * question.answer.length);
    const selectedAnswer = question.answer[randomOption];

    component.find({ value: selectedAnswer }).simulate('change', { target: { value: selectedAnswer } });
    component.update();

    expect(component.state().answer).toBe(selectedAnswer);
  });

  it('should invoque the callback with the selected option', () => {
    const randomOption = Math.floor(Math.random() * question.answer.length);
    const selectedAnswer = question.answer[randomOption];

    component.find({ value: selectedAnswer }).simulate('change', { target: { value: selectedAnswer } });
    component.find('button').simulate('click');

    expect(handleAnswer.mock.calls.length).toBe(1);
    expect(handleAnswer.mock.calls[0][0]).toEqual({ type: question.type, answer: selectedAnswer });
  });

  it('should clear answer from state after execute the callback', () => {
    const newState = Question.getDerivedStateFromProps({ type: 'foo' }, { type: 'bar', answer: 'bar' });

    expect(newState.answer).toBe(null);
  });
});
