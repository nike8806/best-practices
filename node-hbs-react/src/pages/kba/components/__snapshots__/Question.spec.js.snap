// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KBA Question component should render correctly 1`] = `
<div
  className="question"
>
  <h5
    className="question__title todo-heading"
  >
    In which county have you lived?
  </h5>
  <ul
    className="question__answers margin-bottom-32 list-unstyled"
  >
    <li
      className="question__answer"
      key="current.county.b-FULTON"
    >
      <label
        className="question__answer-text"
      >
        <input
          checked={false}
          className="question__answer-selection todo-radio"
          disabled={false}
          name="answer"
          onChange={[Function]}
          type="radio"
          value="FULTON"
        />
        <span
          className="question__answer-radio todo-radio-selector"
        />
        FULTON
      </label>
    </li>
    <li
      className="question__answer"
      key="current.county.b-APPANOOSE"
    >
      <label
        className="question__answer-text"
      >
        <input
          checked={false}
          className="question__answer-selection todo-radio"
          disabled={false}
          name="answer"
          onChange={[Function]}
          type="radio"
          value="APPANOOSE"
        />
        <span
          className="question__answer-radio todo-radio-selector"
        />
        APPANOOSE
      </label>
    </li>
    <li
      className="question__answer"
      key="current.county.b-LAFAYETTE"
    >
      <label
        className="question__answer-text"
      >
        <input
          checked={false}
          className="question__answer-selection todo-radio"
          disabled={false}
          name="answer"
          onChange={[Function]}
          type="radio"
          value="LAFAYETTE"
        />
        <span
          className="question__answer-radio todo-radio-selector"
        />
        LAFAYETTE
      </label>
    </li>
    <li
      className="question__answer"
      key="current.county.b-None of the above"
    >
      <label
        className="question__answer-text"
      >
        <input
          checked={false}
          className="question__answer-selection todo-radio"
          disabled={false}
          name="answer"
          onChange={[Function]}
          type="radio"
          value="None of the above"
        />
        <span
          className="question__answer-radio todo-radio-selector"
        />
        None of the above
      </label>
    </li>
  </ul>
  <button
    className="btn btn-todo center-block question__next"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Next
  </button>
</div>
`;
